### PR TITLE
Restrict arguments of unstack to avoid StackOverflow

### PR DESCRIPTION
**Pull request by bkamins**
Now `unstack` throw StackOverflow if it is passed vector of symbols or numbers. This PR restricts the types of passed arguments only to `Symbol` and `<:Real`.

### CODE REVIEW

File: src/abstractdataframe/reshape.jl [hunk 1]
```diff
@@ -215,7 +215,8 @@ function unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)
     col = similar_missing(df[rowkey], length(levs))
     insert!(payload, 1, copy!(col, levs), _names(df)[rowkey])
 end
-unstack(df::AbstractDataFrame, rowkey, colkey, value) =
+unstack(df::AbstractDataFrame, rowkey::Union{Symbol, Real},
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: You can use the `ColumnIndex` alias.

Comment (bkamins) [reply src/abstractdataframe/reshape.jl]: fixed

**Comment by bkamins** [comment]
Additionally the design of `unstack` now allows:
* single column as `rowkey`;
* all columns except `colkey` and `value` as `rowkey`.

@nalimilan Do you know why we do not allow more than one column but not all columns as `rowkey` (it can be now achieved by subsetting the columns of `DataFrame` and then running `unstack` in the second mode , but it is clumsy)?

**Comment by nalimilan** [comment]
> Additionally the design of unstack now allows:
> 
>     single column as rowkey;
>     all columns except colkey and value as rowkey.
>


> Additionally the design of unstack now allows:
> 
>     single column as rowkey;
>     all columns except colkey and value as rowkey.
>
> @nalimilan Do you know why we do not allow more than one column but not all columns as rowkey (it can be now achieved by subsetting the columns of DataFrame and then running unstack in the second mode , but it is clumsy)?

So you mean it's inconsistent not to allow vectors as `rowkey`, except by implicitly omitting that argument? I'm not sure this is intended.

**Comment by bkamins** [comment]
For now I restrict it.
`colkey`, `value` should always be `ColumnIndex` type.

> So you mean it's inconsistent not to allow vectors as `rowkey`, except by implicitly omitting that argument? 

This is exactly what I mean. I do not have time to work on improving `unstack` to accept vectors in `rowkey` at the moment so I propose to leave the restriction as it stands.

In the future I can rewrite `unstack` to accept vectors in `rowkey` if it is not done earlier by someone else.

**Comment by nalimilan** [comment]
OK, makes sense.

**Comment by bkamins** [comment]
@nalimilan one more small thing fixed in `unstack` with two arguments passed - it was not correctly handling a corner case when there was no possible key column.

File: test/data.jl [hunk 2]
```diff
@@ -189,6 +189,15 @@ module TestData
     @test d1us2[:d] == d1[:d]
     @test d1us2[:3] == d1[:d]
 
+    #
```
Comment (nalimilan) [inline test/data.jl]: ?

File: test/data.jl [hunk 3]
```diff
@@ -189,6 +189,15 @@ module TestData
...
+
+    df1 = melt(DataFrame(rand(10,10)))
+    df1[:id] = 1:100
+    @test size(unstack(df1, :variable, :value)) == (100, 11)
+    # test empty keycol
```
Comment (nalimilan) [inline test/data.jl]: With the line break afterward, it's not clear what line this comment is attached to.

File: test/data.jl (hunk 2 above)
Comment (nalimilan) [inline test/data.jl]: ?

File: test/data.jl (hunk 3 above)
Comment (bkamins) [reply test/data.jl]: I had it fixed - by error I did not push the commit. Should be OK now

**Comment by bkamins** [comment]
The last fix actually showed another problem: `unstack` produced different results when specifying the keycol explicitly (by a name) or implicitly (as all columns except variable and value).

The difference is that when keycol (`rowkey`) was specified unused levels of `CategoricalArray` were reported with `missing` values of respective variables. However, when it was not specified it was dropped (as `groupby` drops it). I recommend to drop it by calling `droplevels!` in https://github.com/JuliaData/DataFrames.jl/pull/1306/files#diff-15c11ad91975e38974d256d28f21db51R197. I understand that it is safe (i.e. the `CategoricalArray` was earlier `deepcopied` and this does not alter the original `rowkey`).

File: src/abstractdataframe/reshape.jl [hunk 4]
```diff
@@ -194,6 +194,7 @@ function unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)
     # `colkey` integer indicating which column to place along column headers
     # `value` integer indicating which column has values
     refkeycol = CategoricalArray{Union{eltype(df[rowkey]), Missing}}(df[rowkey])
+    droplevels!(refkeycol)
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Better add a comment explaining that we do this for consistency with the case when the column is not specified.

Comment (bkamins) [reply src/abstractdataframe/reshape.jl]: fixed

File: test/data.jl [hunk 5]
```diff
@@ -189,6 +189,14 @@ module TestData
     @test d1us2[:d] == d1[:d]
     @test d1us2[:3] == d1[:d]
 
+    # test unstack with exactly one key column that is not passed
+    df1 = melt(DataFrame(rand(10,10)))
+    df1[:id] = 1:100
+    @test size(unstack(df1, :variable, :value)) == (100, 11)
+    # test empty keycol
+
```
[the remark this answers is not in the document]
Comment (bkamins) [reply test/data.jl]: fixed

**Comment by bkamins** [comment]
Maybe the commit should be "Improve unstack: 1) avoid StackOverflow on call, 2) make it consistent when called with and without colkey, 3) improve handling of call without colkey and 0 or 1 possible rowkey variables"

File: src/abstractdataframe/reshape.jl [hunk 6]
```diff
@@ -215,16 +218,20 @@ function unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)
...
-unstack(df::AbstractDataFrame, colkey, value) =
+unstack(df::AbstractDataFrame, colkey::ColumnIndex, value::ColumnIndex) =
     unstack(df, index(df)[colkey], index(df)[value])
 
 function unstack(df::AbstractDataFrame, colkey::Int, value::Int)
     # group on anything not a key or value:
-    g = groupby(df, setdiff(_names(df), _names(df)[[colkey, value]]), sort=true)
+    rowkeys = setdiff(_names(df), _names(df)[[colkey, value]])
+    length(rowkeys) == 0 && throw(ArgumentError("No key column found"))
+    length(rowkeys) == 1 && return unstack(df, rowkeys[1], colkey, value) # better performance
```
Comment (cjprybol) [inline src/abstractdataframe/reshape.jl]: @bkamins do you have the benchmarks you ran to show this has better performance? Just interested in seeing the difference

Comment (bkamins) [reply src/abstractdataframe/reshape.jl]: Thank you for the question. I have profiled it earlier and I thought it gave the benefit. Now - given to report it - actually `groupby` approach is a bit faster.

The cost difference is between line (expensive):
```
i = Int(CategoricalArrays.order(refkeycol.pool)[refkeycol.refs[k]])
```
and line (cheap)
```
i = rowkey[k]
```

I have additionally found a bug in `unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)`:
```
julia> using DataFrames

julia> x = DataFrame(id=[missing,1,2,3], a=1:4, b=1:4)
4×3 DataFrames.DataFrame
│ Row │ id      │ a │ b │
├─────┼─────────┼───┼───┤
│ 1   │ missing │ 1 │ 1 │
│ 2   │ 1       │ 2 │ 2 │
│ 3   │ 2       │ 3 │ 3 │
│ 4   │ 3       │ 4 │ 4 │

julia> y = melt(x, :id)
8×3 DataFrames.DataFrame
│ Row │ variable │ value │ id      │
├─────┼──────────┼───────┼─────────┤
│ 1   │ a        │ 1     │ missing │
│ 2   │ a        │ 2     │ 1       │
│ 3   │ a        │ 3     │ 2       │
│ 4   │ a        │ 4     │ 3       │
│ 5   │ b        │ 1     │ missing │
│ 6   │ b        │ 2     │ 1       │
│ 7   │ b        │ 3     │ 2       │
│ 8   │ b        │ 4     │ 3       │

julia> unstack(y)
ERROR: BoundsError: attempt to access 3-element Array{UInt32,1} at index [0]
Stacktrace:
 [1] getindex(::Array{UInt32,1}, ::UInt32) at .\multidimensional.jl:298
 [2] unstack(::DataFrames.DataFrame, ::Int64, ::Int64, ::Int64) at D:\Software\JULIA_PKG\v0.6\DataFrames\src\abstractdataframe\reshape.jl:205
 [3] unstack(::DataFrames.DataFrame) at D:\Software\JULIA_PKG\v0.6\DataFrames\src\abstractdataframe\reshape.jl:254

julia> unstack(y, :variable, :value)
4×3 DataFrames.DataFrame
│ Row │ id      │ a │ b │
├─────┼─────────┼───┼───┤
│ 1   │ 1       │ 2 │ 2 │
│ 2   │ 2       │ 3 │ 3 │
│ 3   │ 3       │ 4 │ 4 │
│ 4   │ missing │ 1 │ 1 │
```

As @nalimilan recommended - I do not want to push too much into one PR.

So my proposal to you is:
1. merge this PR when you feel it is checked enough
2. I have another PR almost ready that I will put on top of this one that
    * removes current implementation of `unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)`
    * introduces `unstack(df::AbstractDataFrame, rowkey::Vector{Int}, colkey::Int, value::Int)` allowing to index on any columns we want;
    * makes old `unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)` and `unstack(df::AbstractDataFrame, colkey::Int, value::Int)` fall back to the above single function.

OK?

Comment (nalimilan) [reply src/abstractdataframe/reshape.jl]: +1 for another, but if it's not actually faster better remove the comment and the special case from this PR than having incorrect statements in the history.

Comment (bkamins) [reply src/abstractdataframe/reshape.jl]: OK - removed the offending line. I will fix the rest of `unstack` after this is merged.
